activerecord/lib/active_record/connection_adapters/abstract/transaction.rb

```diff
@@ -208,7 +211,16 @@ def current_transaction
       end
 
       private
+
         NULL_TRANSACTION = NullTransaction.new
+
+        # Deallocate invalidated prepared statements outside of the transaction
+        def after_failure_actions(transaction, error)
+          return unless transaction.is_a?(RealTransaction)
+          return unless error.is_a?(ActiveRecord::PreparedStatementCacheExpired)
```

Comment:
I agree in principle, except I couldn't find any place to put adapter-specific transaction behavior.

This method is the 'lowest' possible point in which we can access something outside of the transaction. While it's not ideal, I can't see a less invasive way of implementing this.